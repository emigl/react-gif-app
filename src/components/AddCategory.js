import React, { useState } from 'react'

import PropTypes from 'prop-types'

// Desestructurar las props es mas sencillo que llamar a props.setCategories en este ejemplo
const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('')

    const handleInput = (e) => {
        setInputValue(e.target.value)
        console.log('handleInput triggered')
        
    }

    const handleSubmit = (e) => {
        e.preventDefault()
        console.log('hola', inputValue)
        if( inputValue.trim().length >= 2){
            // la funcion set si no quieres pasar la referencia de los valores del useState de un componente se puede realizar de esta manera, ya que por parametro llama a el valor con un callback y es mÃ¡s sencillo
            setCategories( cats => [ inputValue, ...cats])
            setInputValue('')

        }


    }

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                onChange={handleInput}
                autoFocus
             />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}

export default AddCategory
