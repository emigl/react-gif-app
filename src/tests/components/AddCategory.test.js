import {shallow} from 'enzyme'
import AddCategory from '../../components/AddCategory'
describe('Test para el componente AddCategory', () => {
    // jest.fn() es una función vacía para simular parametros de funcion, pero este te permite varias funcionalidades  de testing extra gracias a jest.
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />)
    const value = 'Mr Robot'
    




    beforeEach(() => {
        // Para limpiar los mocks en cada test.
        jest.clearAllMocks()
        wrapper = shallow(<AddCategory setCategories={setCategories} />)
    })

    test('deberia de renderizar el componente correctamente con snapshot', () => {

        expect(wrapper).toMatchSnapshot()

    })
    
    test('debe de cambiar la caja de texto', () => {


        const input = wrapper.find('input')
        
        // simulate como segundo parametro hay que poner el valor que espera la funcion que se está simulando, con un objeto en este caso.
        input.simulate('change', {
            target: {
                value: value
            }
        })
    })

    test('no debe de postear con el boton submit', () => {
        
        wrapper.find('form').simulate('submit', { preventDefault(){} })

        expect(setCategories).not.toHaveBeenCalled()

    })

    test('debe de llamar el setCategories y limpiar el input', () => {

        const input = wrapper.find('input')

        input.simulate('change', {
            target: {
                value: value
            }
        })

        wrapper.find('form').simulate('submit', {preventDefault(){}})

        // toHaveBeenCalled sirve para ver si se ha llamado a una funcion pasada por el parametro de expect()
        expect(setCategories).toHaveBeenCalled()
        expect(setCategories).toHaveBeenCalledTimes(1)
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function))

        expect(input.text()).toBe('')



    })
    
    
    
})
