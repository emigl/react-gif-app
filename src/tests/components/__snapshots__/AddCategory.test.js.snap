// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test para el componente AddCategory deberia de renderizar el componente correctamente con snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
